### PR TITLE
[WIP] MILESTONE-3: Testnet-Ready

### PR DESCRIPTION
**Pull request by kinosang**
* [x] benchmark
* [x] gensis
* [x] **must increase the `spec_version` to 302 or higher before merge**

### CODE REVIEW

File: pallets/ad/src/lib.rs [hunk 1]
```diff
@@ -468,7 +471,7 @@ pub mod pallet {
                 socring += score;
```
Comment (coderfengyun) [inline pallets/ad/src/lib.rs:471]: saturating_add

File: pallets/did/rpc/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        }
+    }
+}
+
+impl<T, DecentralizedId> DidApi<DecentralizedId> for DidRpcHandler<T, DecentralizedId>
+where
+    T: OffchainStorage + 'static,
+    DecentralizedId: Codec + Send + Sync + 'static,
+{
+    fn get_metadata(&self, did: DecentralizedId) -> Result<(String, String, String)> {
```
Comment (coderfengyun) [inline pallets/did/rpc/src/lib.rs:52]: Can replaced by multi-get keys ?

File: pallets/did/src/lib.rs [hunk 3]
```diff
@@ -251,40 +250,43 @@ pub mod pallet {
...
+        }
 
-            <Metadata<T>>::insert(&did, meta);
+        /// Set bio of a DID.
+        // #[pallet::weight(T::WeightInfo::set_bio(bio.len() as u32))]
+        #[pallet::weight(1_000_000_000)]
+        pub fn set_bio(origin: OriginFor<T>, bio: Vec<u8>) -> DispatchResult {
+            let (did, _) = EnsureDid::<T>::ensure_origin(origin)?;
 
-            Self::deposit_event(Event::<T>::Updated(did));
```
Comment (coderfengyun) [inline pallets/did/src/lib.rs:264]: Won't emit event any more?

File: pallets/did/src/lib.rs [hunk 4]
```diff
@@ -251,40 +250,43 @@ pub mod pallet {
...
+            let key = derive_key(b"bio", &did);
+            sp_io::offchain_index::set(&key, &bio);
 
             Ok(())
         }
 
         /// Set nickname of a DID.
-        #[pallet::weight(T::WeightInfo::set_nickname(nickname.len() as u32))]
+        #[pallet::weight(1_000_000_000)]
```
Comment (coderfengyun) [inline pallets/did/src/lib.rs:276]: magic no. ?

File: pallets/did/src/lib.rs [hunk 5]
```diff
@@ -350,6 +348,10 @@ impl<T: Config> Pallet<T> {
         dest.as_mut().copy_from_slice(&src[(src.len() - len)..]);
         dest
     }
+
+    pub fn zero() -> T::DecentralizedId {
```
Comment (coderfengyun) [inline pallets/did/src/lib.rs:352]: zero -> default?

File: pallets/did/src/tests.rs [hunk 6]
```diff
@@ -164,64 +148,78 @@ fn should_fail_when_already_have_did() {
...
 
         assert_noop!(
-            Did::transfer(Origin::signed(alice), bob),
+            Did::transfer(Origin::signed(ALICE), BOB),
             Error::<Test>::NotExists
         );
     });
 }
 
 #[test]
```
Comment (coderfengyun) [inline pallets/did/src/tests.rs:160]: For lib.rs, tests can be arranged into multiple modules, let them can be init independently.

File: pallets/did/src/tests.rs [hunk 7]
```diff
@@ -164,64 +148,78 @@ fn should_fail_when_already_have_did() {
...
 fn should_set_avatar() {
-    new_test_ext().execute_with(|| {
-        let alice = sr25519::Public([1; 32]);
-        let did = DID::from_slice(&[0xff; 20]);
+    let db = TestPersistentOffchainDB::new();
```
Comment (coderfengyun) [inline pallets/did/src/tests.rs:162]: If define meta_tests module, these repeated init logic can be avoid.
Or
can extract these code as fun.

File: pallets/did/src/tests.rs [hunk 8]
```diff
@@ -164,64 +148,78 @@ fn should_fail_when_already_have_did() {
...
-        let meta = <Metadata<Test>>::get(did).unwrap();
-        assert_eq!(meta.avatar, vec![1, 2, 3]);
+    let db = t.offchain_db();
+    let avatar = db.get(&derive_key(b"avatar", &DID_ALICE));
+    assert_eq!(avatar, Some(vec![1, 2, 3]));
+}
+
+#[test]
+fn should_set_bio() {
+    let db = TestPersistentOffchainDB::new();
```
Comment (coderfengyun) [inline pallets/did/src/tests.rs:180]: If define meta_tests module, these repeated init logic can be avoid.
Or
can extract these code as fun.

File: pallets/did/src/tests.rs [hunk 9]
```diff
@@ -164,64 +148,78 @@ fn should_fail_when_already_have_did() {
...
+        assert_ok!(Did::set_avatar(Origin::signed(ALICE), vec![1, 2, 3]));
+    });
 
-        assert_ok!(Did::set_avatar(Origin::signed(alice), vec![1, 2, 3]));
+    t.persist_offchain_overlay();
 
-        let meta = <Metadata<Test>>::get(did).unwrap();
-        assert_eq!(meta.avatar, vec![1, 2, 3]);
+    let db = t.offchain_db();
+    let avatar = db.get(&derive_key(b"avatar", &DID_ALICE));
```
Comment (coderfengyun) [inline pallets/did/src/tests.rs:174]: There prefixes should be defined as const variables.

File: pallets/did/src/tests.rs [hunk 10]
```diff
@@ -164,64 +148,78 @@ fn should_fail_when_already_have_did() {
...
+    t.register_extension(OffchainDbExt::new(offchain.clone()));
+
+    t.execute_with(|| {
+        assert_ok!(Did::set_bio(Origin::signed(ALICE), vec![1, 2, 3]));
     });
+
+    t.persist_offchain_overlay();
+
+    let db = t.offchain_db();
+    let avatar = db.get(&derive_key(b"bio", &DID_ALICE));
```
Comment (coderfengyun) [inline pallets/did/src/tests.rs:192]: There prefixes should be defined as const variables.

File: pallets/did/src/tests.rs [hunk 11]
```diff
@@ -164,64 +148,78 @@ fn should_fail_when_already_have_did() {
...
+    });
 
-        assert_ok!(Did::set_nickname(Origin::signed(alice), vec![1, 2, 3]));
+    t.persist_offchain_overlay();
 
-        let meta = <Metadata<Test>>::get(did).unwrap();
-        assert_eq!(meta.nickname, vec![1, 2, 3]);
-    });
+    let db = t.offchain_db();
+    let avatar = db.get(&derive_key(b"nickname", &DID_ALICE));
```
Comment (coderfengyun) [inline pallets/did/src/tests.rs:210]: There prefixes should be defined as const variables.

File: pallets/did/utils/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,28 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Codec;
+use parami_primitives::names;
+use sp_std::prelude::*;
+
+pub fn derive_key<C: Codec>(group: &[u8], key: &C) -> Vec<u8> {
```
Comment (coderfengyun) [inline pallets/did/utils/src/lib.rs:7]: serialize_to_store_key

File: pallets/magic/src/lib.rs [hunk 13]
```diff
@@ -111,7 +111,7 @@ pub mod pallet {
         pub fn create_stable_account(
             origin: OriginFor<T>,
             magic_account: AccountOf<T>,
-            #[pallet::compact] deposit: BalanceOf<T>,
+            #[pallet::compact] stashed: BalanceOf<T>,
```
Comment (coderfengyun) [inline pallets/magic/src/lib.rs:114]: What's the diff between 'deposit' and 'stashed' ?

File: pallets/nft/src/lib.rs [hunk 14]
```diff
@@ -416,14 +422,22 @@ impl<T: Config> Pallet<T> {
...
-                maximum,
-                now,
                 started,
+                maximum,
+                height,
+                supply,
             );
 
-            let amount = if amount < maximum { amount } else { maximum };
+            let left = maximum - supply;
```
Comment (coderfengyun) [inline pallets/nft/src/lib.rs:435]: totalSupplied ?

File: pallets/nft/src/lib.rs [hunk 15]
```diff
@@ -416,14 +422,22 @@ impl<T: Config> Pallet<T> {
...
+            let amount = if amount < left { amount } else { left };
```
Comment (coderfengyun) [inline pallets/nft/src/lib.rs:436]: Is there Math.min alike func ?

File: runtime/src/lib.rs [hunk 16]
```diff
@@ -1150,21 +1151,21 @@ impl parami_xassets::Config for Runtime {
...
-    pub const DidPalletId: PalletId = PalletId(*b"prm/did ");
+    pub const DidPalletId: PalletId = PalletId(*names::DID);
 }
 
 impl parami_did::Config for Runtime {
     type Event = Event;
     type AssetId = AssetId;
     type Currency = Balances;
-    type DecentralizedId = sp_core::H160;
+    type DecentralizedId = DecentralizedId;
```
Comment (coderfengyun) [inline runtime/src/lib.rs:1161]: Why need to redeclare this type ?

File: pallets/chainbridge/src/lib.rs [hunk 17]
```diff
@@ -6,11 +6,11 @@ pub use types::*;
 #[rustfmt::skip]
 pub mod weights;
 
-#[cfg(test)]
-mod mock;
+// #[cfg(test)]
```
Comment (coderfengyun) [inline pallets/chainbridge/src/lib.rs:9]: Shall we just comment out the specific case, not the whole suite instead ?

File: pallets/xassets/src/lib.rs [hunk 18]
```diff
@@ -5,11 +5,11 @@ pub use pallet::*;
 #[rustfmt::skip]
 pub mod weights;
 
-#[cfg(test)]
-mod mock;
+// #[cfg(test)]
```
Comment (coderfengyun) [inline pallets/xassets/src/lib.rs:8]: Shall we comment out chain-bridge's specific can't-pass case, not the whole suite instead ?

File: pallets/did/src/lib.rs [hunk 19]
```diff
@@ -251,40 +250,43 @@ pub mod pallet {
         }
 
         /// Set avatar of a DID.
-        #[pallet::weight(T::WeightInfo::set_avatar(avatar.len() as u32))]
+        #[pallet::weight(1_000_000_000)]
         pub fn set_avatar(origin: OriginFor<T>, avatar: Vec<u8>) -> DispatchResult {
```
[the remark this answers is not in the document]
Comment (kinosang) [reply pallets/did/src/lib.rs:254]: wait benchmarking

File: pallets/did/src/lib.rs (hunk 5 above)
Comment (kinosang) [reply pallets/did/src/lib.rs:352]: `default` is reserved for rust `Default` trait

File: pallets/magic/src/lib.rs (hunk 13 above)
Comment (kinosang) [reply pallets/magic/src/lib.rs:114]: `stashed` is more precise to this value. it's the amount to be transfer to stash account.

File: pallets/nft/src/lib.rs (hunk 15 above)
Comment (kinosang) [reply pallets/nft/src/lib.rs:448]: no

File: runtime/src/lib.rs (hunk 16 above)
Comment (kinosang) [reply runtime/src/lib.rs:1176]: for pallet coupling

File: pallets/chainbridge/src/lib.rs (hunk 17 above)
Comment (kinosang) [reply pallets/chainbridge/src/lib.rs:9]: no, mock and tests from ChainBridge pallets failed to build

**Pull request merged**